tmk_core/protocol/chibios/usb_driver.c

```diff
@@ -154,312 +78,254 @@ static void obnotify(io_buffers_queue_t *bqp) {
+    if (size > 0) {
+        /* Posting the filled buffer in the queue.*/
+        ibqPostFullBufferI(&endpoint->ibqueue, usbGetReceiveTransactionSizeX(endpoint->config.usbp, endpoint->config.ep));
+    }
 
-    /* The endpoint cannot be busy, we are in the context of the callback,
-       so a packet is in the buffer for sure. Trying to get a free buffer
-       for the next transaction.*/
-    (void)qmkusb_start_receive(qmkusbp);
+    /* The endpoint cannot be busy, we are in the context of the callback, so a
+     * packet is in the buffer for sure. Trying to get a free buffer for the
+     * next transaction.*/
+    usb_start_receive(endpoint);
 
     osalSysUnlockFromISR();
 }
 
-/**
- * @brief   Default data received callback.
- * @details The application must use this function as callback for the IN
- *          interrupt endpoint.
- *
- * @param[in] usbp      pointer to the @p USBDriver object
- * @param[in] ep        endpoint number
- */
-void qmkusbInterruptTransmitted(USBDriver *usbp, usbep_t ep) {
-    (void)usbp;
-    (void)ep;
+bool usb_endpoint_in_send(usb_endpoint_in_t *endpoint, const uint8_t *data, size_t size, sysinterval_t timeout, bool buffered) {
+    osalDbgCheck((endpoint != NULL) && (data != NULL) && (size > 0U) && (size <= endpoint->config.buffer_size));
+
+    osalSysLock();
+    if (usbGetDriverStateI(endpoint->config.usbp) != USB_ACTIVE) {
+        osalSysUnlock();
+        return false;
+    }
+
+    /* Short circuit the waiting if this endpoint timed out before, e.g. if
+     * nobody is listening on this endpoint (is disconnected) such as
+     * `hid_listen`/`qmk console` or we are in an environment with a very
+     * restricted USB stack. The reason is to not introduce micro lock-ups if
+     * the report is send periodically. */
+    if (endpoint->timed_out && timeout != TIME_INFINITE) {
+        timeout = TIME_IMMEDIATE;
+    }
+    osalSysUnlock();
+
+    while (true) {
+        size_t sent = obqWriteTimeout(&endpoint->obqueue, data, size, timeout);
+
+        if (sent < size) {
+            osalSysLock();
+            endpoint->timed_out |= sent == 0;
+            bqSuspendI(&endpoint->obqueue);
+            obqResetI(&endpoint->obqueue);
+            bqResumeX(&endpoint->obqueue);
+            osalOsRescheduleS();
+            osalSysUnlock();
+            continue;
+        }
+
+        if (!buffered) {
+            obqFlush(&endpoint->obqueue);
+        }
+
+        return true;
+    }
 }
 
-/** @} */
+void usb_endpoint_in_flush(usb_endpoint_in_t *endpoint, bool padded) {
+    osalDbgCheck(endpoint != NULL);
+
+    size_t count;
+    if (padded && obqGetFullBufferI(&endpoint->obqueue, &count) != NULL) {
+        ptrdiff_t padding_len = endpoint->config.buffer_size - count;
+        while (padding_len > 0) {
+            obqPutTimeout(&endpoint->obqueue, 0, TIME_IMMEDIATE);
+            padding_len--;
+        }
+    }
+    obqFlush(&endpoint->obqueue);
+}
```

Comment:
This implementation does not look correct — `obqGetFullBufferI()` takes the buffer from the reader end of the queue, not the writer end.

Looks like ChibiOS does not have any API that would be suitable here, so you may try accessing `->ptr` and `->top` directly to determine how many bytes of padding are needed.

Alternatively, you may leave the buffers partially filled here, and pad the reports in the code which actually transmits USB packets (the buffer returned by `obqGetFullBufferI()` actually has the full size allocated, so you may fill the padding there).

Reply:
Good catch! Thanks. I've updated to code to calculate the required padding bytes using `ptr` and `top`.